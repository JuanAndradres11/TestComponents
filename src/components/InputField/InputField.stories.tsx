// src/components/InputFieldComponent.stories.tsx
import React, { useState } from 'react';
import { type Meta, type StoryObj } from '@storybook/react';
import InputFieldComponent from './InputField';

const meta: Meta<typeof InputFieldComponent> = {
  title: 'Components/InputField',
  component: InputFieldComponent,
  tags: ['autodocs'],
};

export default meta;
type Story = StoryObj<typeof InputFieldComponent>;

// Default interactive input
export const Default: Story = {
  render: () => {
    const [value, setValue] = useState('');
    return (
      <InputFieldComponent
        placeholder="Enter your text"
        value={value}
        onChange={(e) => setValue(e.target.value)}
      />
    );
  },
};

// Input with live error message
export const WithValidation: Story = {
  render: () => {
    const [value, setValue] = useState('');
    const isError = value.length > 0 && value.length < 3;
    return (
      <InputFieldComponent
        placeholder="Enter at least 3 characters"
        value={value}
        onChange={(e) => setValue(e.target.value)}
        error={isError ? 'Minimum 3 characters required' : ''}
      />
    );
  },
};

// Pre-filled input field
export const Prefilled: Story = {
  render: () => {
    const [value, setValue] = useState('Hello World');
    return (
      <InputFieldComponent
        placeholder="Editable text"
        value={value}
        onChange={(e) => setValue(e.target.value)}
      />
    );
  },
};
