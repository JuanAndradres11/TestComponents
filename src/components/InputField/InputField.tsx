import React, { useState } from "react";
import {
  TextField,
  InputAdornment,
  IconButton,
  CircularProgress,
} from "@mui/material";
import { Visibility, VisibilityOff, Clear } from "@mui/icons-material";

export interface InputFieldProps {
  value?: string;
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
  label?: string;
  placeholder?: string;
  helperText?: string;
  errorMessage?: string;
  disabled?: boolean;
  invalid?: boolean;
  loading?: boolean;
  variant?: "filled" | "outlined" | "standard" | "ghost"; // added ghost separately
  size?: "sm" | "md" | "lg";
  type?: "text" | "password";
  clearable?: boolean;
}

const InputField: React.FC<InputFieldProps> = ({
  value,
  onChange,
  label,
  placeholder,
  helperText,
  errorMessage,
  disabled = false,
  invalid = false,
  loading = false,
  variant = "outlined",
  size = "md",
  type = "text",
  clearable = false,
}) => {
  const [showPassword, setShowPassword] = useState(false);

  // map sizes
  const sizeMap = {
    sm: "small",
    md: "medium",
    lg: "medium", // MUI doesn't have lg, so we can style later
  } as const;

  // ghost variant handling (simulate by removing borders)
  const isGhost = variant === "ghost";

  return (
    <TextField
      label={label}
      placeholder={placeholder}
      value={value}
      onChange={onChange}
      disabled={disabled || loading}
      error={invalid}
      helperText={invalid ? errorMessage : helperText}
      variant={isGhost ? "standard" : variant} // ghost handled separately
      size={sizeMap[size]}
      type={type === "password" && !showPassword ? "password" : "text"}
      fullWidth
      InputProps={{
        sx: isGhost
          ? {
              border: "none",
              boxShadow: "none",
              "& .MuiInputBase-input": { padding: "6px 0" },
            }
          : {},
        endAdornment: (
          <>
            {loading && (
              <InputAdornment position="end">
                <CircularProgress size={20} />
              </InputAdornment>
            )}
            {clearable && value && (
              <InputAdornment position="end">
                <IconButton onClick={() => onChange?.({ 
                  target: { value: "" } 
                } as React.ChangeEvent<HTMLInputElement>)}>
                  <Clear fontSize="small" />
                </IconButton>
              </InputAdornment>
            )}
            {type === "password" && (
              <InputAdornment position="end">
                <IconButton onClick={() => setShowPassword(!showPassword)}>
                  {showPassword ? <VisibilityOff /> : <Visibility />}
                </IconButton>
              </InputAdornment>
            )}
          </>
        ),
      }}
    />
  );
};

export default InputField;
