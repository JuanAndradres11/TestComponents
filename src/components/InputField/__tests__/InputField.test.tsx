import { render, screen, fireEvent } from "@testing-library/react";
import InputField from "../InputField";

describe("InputField", () => {
  it("renders with placeholder", () => {
    render(<InputField placeholder="Enter text" />);
    expect(screen.getByPlaceholderText("Enter text")).toBeInTheDocument();
  });

  it("updates value when typing", () => {
    render(<InputField placeholder="Type here" />);
    const input = screen.getByPlaceholderText("Type here") as HTMLInputElement;
    fireEvent.change(input, { target: { value: "hello" } });
    expect(input.value).toBe("hello");
  });

  it("respects disabled prop", () => {
    render(<InputField placeholder="Disabled" disabled />);
    const input = screen.getByPlaceholderText("Disabled");
    expect(input).toBeDisabled();
  });
});
