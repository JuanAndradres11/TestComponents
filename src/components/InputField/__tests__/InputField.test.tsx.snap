// Jest Snapshot v1, for InputField.test.tsx
exports[`Input field updates correctly 1`] = `
<input placeholder="Search..." value="Test" />
`;
